### ISSUE TITLE
Non-blocking API for long-running operations

### ISSUE THREAD

**Issue by athoelke** [ARM-software/psa-firmware-update-spec#17]
### Use case

There is demand for the PSA Firmware Update API to enable the following use case:

* To remain operational during the firmware update process, my application requires that calls to the PSA Firmware Update API by the Update Client have a bounded latency.

This availability requirement for the application is typically not in force during a reboot, and any installation activity within the Bootloader has no effect on the this API design.

This is particularly relevant for constrained systems that do not provide full multi-tasking, where a blocking call in the Update Client will block the whole application from responding to latency-sensitive events.

The implication for the implementation and the API depends on the maximum permitted time for a call, the hardware design and type of non-volatile memories (NVM), and the phases of the update process for which the requirement holds.

### Analysis

As a persistent state model, and with APIs specifically defined to modify the data in the Firmware Store, most of the operations in the PSA Firmware Update API will result in write and erase actions on the NVM. Depending on the NVM technology, write and erase operations can take from <1us to 10s of ms. In particular, flash memories in current use can have erase operations that apply to large blocks of NVM and take significant time to complete (>>1ms). Raw writes to flash NVMs are typically fast, but if writes are paged, this results in read/modify/write of a page of NVM, and will occasionally result in an erase cycle to refresh some of the storage.

In addition, these operations on NVM are typically implemented independently within the NVM hardware. The operation does not require the MCU to actively drive the operations, so the MCU could continue other execution while the NVM operation happens asynchronously.

A simple first objective is to enable the Client to carry out the cleaning/erasing of the *second* image in an asynchronous or step-wise manner, where each call to the `clean` API does not block for a long time, and can return before the operation completes. This might be sufficient for applications which does not require writes to NVM to be non-blocking. The Client can already control the size of data that is written when transferring the firmware image to the Update Service.

The state model defined in #1, deliberately isolates most of the NVM erase activity to `clean` operation. Although this could be deferred until `start`, expecting this activity to take place during the `clean` operation might enable a simpler API that can support non-blocking erase. 

A much more complex goal is to allow operations that write to flash to return before completing.

A more complete discussion of the challenges and possible solutions follows after the outline proposal.

### Proposal

Enable erase operations to be implemented in an incremental or asynchronous manner. This would be achieved by:

* Adding a CLEANING state, in which the *second* image contains the remnants of a previous update or attempted update.
* Any `clean` operation call, starting from WRITING, STAGED, FAILED, or UPDATED, might immediately reach the READY state, or can result in the CLEANING state if the process is incomplete.
* In the CLEANING state, the only valid operation is `clean`, which will continue the process. If complete, this transitions to READY, otherwise the Store remains in the CLEANING state.
* The CLEANING state is optionally persistent - an implementation can complete a started `clean` operation during boot, before the Client can query the state of the Component; or it can report the CLEANING state to the Client after a reboot.

The state model would be as follows:

![image](https://user-images.githubusercontent.com/5690199/172743239-feb52cd8-cb35-4dfb-b4df-2c0888680a25.png)

### Challenge 1: Intermediate states

Because the Firmware Store has a persistent state model, intermediate states during long-running operations can be visible to the Client. This requires additional states to be added to the model.

#### Option A: Hide the intermediates

This could be done by one or more of the following techniques:
* Refuse to report on the Firmware Store status while a long-running operation is ongoing (e.g. return a BUSY error).
* Report the starting state for the transition while a long-running operation is ongoing.
* Automatically continue and finish, or fail, the operation on a restart before the Client can query the state.

#### Option B: Add states to the state model, and define the transitions

This makes the state model appear more complex, but results in less implementation-specific behavior relating to this support.

#### Proposal

Design the API around option B: the intermediate states required to manage non-atomic operations make the implementation and client more complex. For correct and secure use of an API, it is preferable to make that complexity explicit. 

### Challenge 2: Consuming input data

If an operation that takes data as input can return before the operation completes, what is the status of the data buffer provided by the Update Client between the return from the call, and the completion of the entire operation? - for example, if we want to make `write` a non-blocking operation.

#### Option A: Defer

Don't tackle a non-blocking `start` or `write` operation until after v1.0.

#### Option B: The Service retains access to the buffer until the operation completes

In this approach, the buffer could be handed directly over to hardware before returning from the call. The Client is required to retain the buffer (unmodified) until the operation is reported to be complete.

#### Option C: Consume the provided data before the call returns

In this approach, the Service must make a copy of any provided data that it does not use immediately. The Client is free to reuse or recycle the memory used for the input parameter, as soon as the function call returns. The Client does not need to wait for the operation to finish.

#### Proposal

Either option A for v1.0 (leave this out of scope) or option C.

The Client not being able to safely manage the buffer after the function returns is non-intuitive: shared memory buffers are a reliability/correctness hazard in software systems. For Update Service implementations that are within a Root of Trust, it would be recommended to copy the data into memory controlled by the Service, before sending the data to NVM.

### Challenge 3: Aborting a non-blocking operation

If an operation is in progress, and the Client then determines that an alternative course of action is required, is there a need to provide 'cancel' functionality to try and terminate/abort the ongoing operation as early as possible? If a cancel function is provided, it should be atomic, and only return when the operation has completed or been aborted - any resources used by the in-progress operation must be released.

If all intermediate states are defined in the API, this would take the form of a permitted transition from the intermediate state to a state other than the original target state.

**Comment by athoelke** [comment]
## API design approaches

### A: Just use a return value to indicate unfinished operations?

One simple approach is to have an existing API function return a new not-error status code to indicate that the operation is still in progress. The Client must call the API again 'some time later' (decided by the Client) to determine if the operation is still 'in progress' or if it has completed (successfully or with an error).

To illustrate this idea, suppose we add `PSA_FWU_INCOMPLETE` as a return code to indicate an in-progress `clean` operation for `psa_fwu_clean()`. A client could use this API to yield to other tasks between steps of the cleaning operation:

``` c
psa_status_t r;
while ((r = psa_fwu_clean(image)) == PSA_FWU_INCOMPLETE) {
    task_yield();
}
if (r != PSA_SUCCESS) {
    // handle error
}
```

This would work with implementations that were blocking, and with implementations that return early.

However, this is an incompatible change to the function behavior. A correct client using the `clean` API before it can be non-blocking would be:

``` c
psa_status_t r = psa_fwu_clean(image);
if (r != PSA_SUCCESS) {
    // handle error
}
```

After the API becomes non blocking, this is no longer correct code. It would have to change, at least, to the following:

``` c
psa_status_t r;
while ((r = psa_fwu_clean(image)) == PSA_FWU_INCOMPLETE)
   ;
if (r != PSA_SUCCESS) {
    // handle error
}
```

This means that we cannot introduce this to the API gradually, without changing the major version, and creating unhelpful churn and adding risk to existing clients at that time.

Another negative consequence of this approach, is that when a client does not need to have the operation complete early, it must still write the looping code, as it cannot indicate to the implementation that it will wait. This will often be less energy efficient than allowing the implementation to wait for the operation to complete.

Finally, the original non-looping code for this operation still *looks* correct, and the error is less likely to be detected in a code review or during code maintenance.

### B: Use a configuration API to select non-blocking operation

This looks mostly like option (A) above, but operations will default to blocking behavior. The client can enable non-blocking behavior by using a new API. For example:

``` c
psa_status_t r = psa_fwu_set_nonblocking(PSA_FWU_NONBLOCKING_ENABLE);
```

This appears to solve the compatibility issue with (A), when introducing non-blocking behavior later on. The configuration API would have to be complex enough to allow independent control for APIs that were made non-blocking at different times.

Specifying client requested behavior via configuration is more complex than via an explicit API, and also makes code review and maintenance more difficult as the behavior of a call to `psa_fwu_clean()` (and therefore the correct form of usage) is controlled by a possibly distant call to the configuration API.

### C: Add new APIs for non-blocking operations

For example, for `psa_fwu_clean()` we add `psa_fwu_clean_nowait()`, which would be used like this:

``` c
psa_status_t r;
while ((r = psa_fwu_clean_nowait(image)) == PSA_FWU_INCOMPLETE) {
    task_yield();
}
if (r != PSA_SUCCESS) {
    // handle error
}
```

An alternative would be to add a parameter to each call which must be `PSA_FWU_WAIT` or `PSA_FWU_NOWAIT`.

This adds the most complexity to the API definition:
* Need to identify a good naming pattern for the non-blocking variants of operations
* Requires specification of optionality of the API: can the functions not be present (build-time failure), or must they return an NOT_SUPPORTED error (run-time failure)

However, this approach enables the caller to be explicit about blocking behavior in each operation call. This has the following beneficial effects:
* Reduces the risk of undetected errors in code
* Does not create a compatibility problem when non-blocking operations are introduced
* Enables efficient implementation when waiting for operations to complete

**Comment by athoelke** [comment]
## Asynchronous versus Incremental implementation

In operation implementations where the majority of the work is done in other hardware (e.g. a Flash memory controller), we can describe the operation as 'asynchronous'. If the system design cannot schedule other work in the middle of a PSA FWU call, then returning early enables the MCU to do useful work while the operation happens concurrently in the peripheral. If there is no other work to be done, it might be better to wait on an interrupt, or sleep for a short time before checking the status of the operation with another call to the service. Otherwise the MCU can be busy in a short polling loop, burning energy for little benefit.

In operation implementations where the majority of the work is done on the MCU, an 'incremental' or 'stepwise' implementation that breaks up the work into smaller pieces can limit the time spent doing work in the call. If the system design cannot schedule other work in the middle of a PSA FWU call, then returning early allows the caller to yield execution so that important tasks can run with bounded latency. If there is no other work to do, then the operation should be continued immediately, as there is no hardware to wait for.

_**Question:**_ Can we have a single non-blocking API for operations that works well for both forms of implementation?

* An asynchronous implementation burns unnecessary energy if the API is called in a tight loop.
* An incremental implementation takes unnecessary time if the API is called with a sleep() delay between calls in a loop.

To optimize both time and energy usage, while also providing bounded latency for API calls - one needs to tune the ideal wait time between calls to a non-blocking operation function. This will either need more API complexity, or we must expect that Clients will be somewhat tailored for the specific Service implementation if this is a critical issue for the product.

**Comment by d3zd3z** [comment]
So far, at least, very few of the MCUs I've used support returning to executing code during flash operations. Those that do, generally divide the flash memory into a fixed set of partitions, and only allow simultaneous operation across different segments.

I agree that supporting this could be useful in the API, but I'm wondering how many devices would actually offer this support.

**Comment by athoelke** [comment]
Based on the state model in that comment, we could just use the FAILED and UPDATED states as the intermediate states for an incremental `clean` operation, as shown in this state diagram:

![image](https://user-images.githubusercontent.com/5690199/177127427-f6f1f564-3079-4a3a-b7c6-0da204983c70.png)

**Comment by athoelke** [comment]
With the state model adjustment to use `reject` to cancel an update operation that has been started, the addition of a new API to provide an incremental `clean` function has a small effect on the overall API. It can be added without changes to the component states and transitions, or to other functions - and so could be introduced in minor (compatible) update in future.

### Proposal

Leave an incremental `clean` function **out** of the first draft for version 1.0. Adding this functionality as a late addition before finalising v1.0, or as a feature for v1.1 will be straight-forward, given the other changes proposed for the v1.0 state model.

**Comment by JohnThorne** [comment]
My comment from the previous meet is that the 'clean/clean_step' operation suites our own implementation better, especially when the image store can store multiple components (e.g. an external NV memory device). Here 'clean' operation simply marks the rejected component as 'dirty' and the 'dirty' component storage area cannot be used until 'cleaned', which usually happens in the 'UPDATED' state (although it can happen pre-READY in the boot process). For us, this removes the need to fully 'clean' the storage results during the active server-client connection (resource costly), and any 'reject' operation during 'WRITING' state results in the component storage operation (by return to 'WRITING' state) to be re-started much quicker.

**Comment by athoelke** [comment]
@JohnThorne - thanks for the comment. I'd like to be sure I have understood the suggestions/request that you are making.

The intention with the states and transitions, is that `start`/`write`/`reject` do not involve expensive flash erase activity, the duration of these calls would be relative to the amount of data being written (e.g. manifest or image data). If an implementation needs to use expensive erase operations, these should be part of the `clean` call OR handled by the implementation during startup/shutdown (and so not impacting a Client call). The latter is easier to provide if the firmware store can either overwrite dirty images without an intervening erase, or has enough room for multiple staged images, leaving dirty images in place until a point when the erase is acceptable.

`clean_step` would replace the use of `clean` for clients/implementations for which `clean` is expensive, and the Client can reduce the blocking time of calls by making multiple calls to `clean_step` instead.

* Are your requirements compatible with this state model?
* Do you need `clean_step` as well as, or instead of, `clean`?
* Is the introduction of `reject` to cancel an update beneficial, problematic, or neutral, for your systems?

**Comment by JohnThorne** [comment]
Hi Andrew,

Thanks for your email. To try to answer your questions…


  *   Are your requirements compatible with this state model?

Yes, I believe they are. The reason I commented was that rather than a true clean (as in attempting a physical erase or instigating a storage memory reformat process), we simply mark the image store area as ‘dirty’ (which is very quick), rather than attempting a physical clean process during the server/client call. The state can then immediately return to READY (if the reason for the reject allowed), and the process restarted, if required. If this is process covered by clean in the specification, then we have no problem with it being called clean in the state machine.

***@***.***

I have also assumed that, by the state retuning to READY following a clean, that this inherently allows the code to immediately return to the WRITING state.


  *   Do you need clean_step as well as, or instead of, clean?

I believe we do not require clean_step. Sorry if I made this unclear by grouping the clean_step and clean transitions in my comment. Having said that, my mind is still exploring possible adaptation of a non-blocking clean_step process being available and if that could be of benefit, but nothing yet.


  *   Is the introduction of reject to cancel an update beneficial, problematic, or neutral, for your systems?

Having a reject transition is how our code generically wants to work, so it is greatly beneficial.

I hope this clarifies things. Shout if it doesn’t.

Best Regards

John Thorne
_______________________________________________________________________

John Thorne - Director
Cambridge Micro Engineering Ltd.
83 High Street
Linton
Cambridge CB21 4JT
UK
Office: +44 (0)1223 893 872
Mobile: +44 (0) 7367 963 059
Email: ***@***.***>
Web Site: <http://www.cme.co.uk<http://www.cme.co.uk/>>
LinkedIn: <johngthorne<https://uk.linkedin.com/in/johngthorne>>
Facebook: ***@***.***><https://www.facebook.com/Cambridge-Micro-Engineering-582479018916727/?modal=admin_todo_tour>
Twitter: ***@***.***><https://twitter.com/CambridgeMicro>

Electronics design consultancy and regulatory compliance engineering
________________________________________________________________________
